src/value/ser.rs

```diff
@@ -348,34 +377,216 @@ impl serde::ser::SerializeMap for SerializeMap {
     where
         T: Serialize,
     {
-        match try!(to_value(&key)) {
-            Value::String(s) => self.next_key = Some(s),
-            Value::Number(n) => {
-                if n.is_u64() || n.is_i64() {
-                    self.next_key = Some(n.to_string())
-                } else {
-                    return Err(Error::syntax(ErrorCode::KeyMustBeAString, 0, 0));
-                }
-            }
-            _ => return Err(Error::syntax(ErrorCode::KeyMustBeAString, 0, 0)),
-        };
-        Ok(())
+        match *self {
+            SerializeMap::Map { ref mut next_key, .. } => {
+                *next_key = Some(try!(key.serialize(MapKeySerializer)));
+                Ok(())
+            },
+            #[cfg(feature = "arbitrary_precision")]
+            SerializeMap::Number { .. } => panic!(),
```

Comment:
Is this reachable? A malicious input should not be able to trigger a panic.

Reply:
No, it's not reachable AFAIK. I should probably change to `unreachable!()`.